src/Storage/Controllers/ContentOnDemandController.cs

```diff
@@ -122,7 +124,61 @@ public async Task<Stream> GetPaymentAsHtml([FromRoute] string org, [FromRoute] s
+            string htmlFormId = htmlElement.Metadata.First(m => m.Key == "formid").Value;
+            DataElement xmlElement = instance.Data.First(d => d.Metadata.First(m => m.Key == "formid").Value == htmlFormId && d.Id != htmlElement.Id);
+            string visiblePagesString = xmlElement.Metadata.FirstOrDefault(m => m.Key == "A2VisiblePages")?.Value;
+            List<int> visiblePages = !string.IsNullOrEmpty(visiblePagesString) ? visiblePagesString.Split(';').Select(int.Parse).ToList() : null;
+
+            int lformid = int.Parse(xmlElement.Metadata.First(m => m.Key == "lformid").Value);
+            PrintViewXslBEList xsls = [];
+            int pageNumber = 1;
+            foreach ((string xsl, bool isPortrait) in await _a2Repository.GetXsls(org, app, lformid, language, 3))
+            {
+                if (visiblePages == null || visiblePages.Contains(pageNumber))
+                {
+                    xsls.Add(new PrintViewXslBE() { PrintViewXsl = xsl, Id = $"{lformid}-{pageNumber}{language}", IsPortrait = isPortrait, PageNumber = pageNumber });
+                }
+
+                ++pageNumber;
+            }
+
+            xsls[^1].LastPage = true;
+
+            if (xsls.Count > 1 && xsls.Exists(x => x.IsPortrait) && xsls.Exists(x => !x.IsPortrait))
+            {
+                // Mix of portrait and landscape, we must generate each page and merge them
+                using (var mergedDoc = new PdfDocument())
+                {
+                    foreach (var xsl in xsls)
+                    {
+                        var pdfPages = await _pdfGeneratorClient.GeneratePdf(
+                            $"{Request.Scheme}://{Request.Host}{Request.PathBase}{Request.Path}{Request.QueryString}"
+                                .Replace("formdatapdf", $"formdatahtml/{xsl.PageNumber}"),
+                            xsl.IsPortrait);
+                        using (var pageDoc = PdfReader.Open(pdfPages, PdfDocumentOpenMode.Import))
+                        {
+                            for (var i = 0; i < pageDoc.PageCount; i++)
+                            {
+                                pageDoc.Pages[i].Orientation = xsl.IsPortrait ? PdfSharp.PageOrientation.Portrait : PdfSharp.PageOrientation.Landscape;
+                                mergedDoc.AddPage(pageDoc.Pages[i]);
+                            }
+                        }
+                    }
+
+                    MemoryStream mergedPdf = new MemoryStream();
+                    mergedDoc.Save(mergedPdf);
+                    return mergedPdf;
+                }
+            }
+            else
+            {
+                // Generate all pages in a single operation
+                return await _pdfGeneratorClient.GeneratePdf(
+                    $"{Request.Scheme}://{Request.Host}{Request.PathBase}{Request.Path}{Request.QueryString}".Replace("formdatapdf", "formdatahtml"),
```

Comment:
Could it be possible for a crafty hacker to make a specially made request that we now relatively blindly forward to puppeteer? Any dangers with this and should we build the URL based on our knowledge of the path and configuration instead?

Reply:
The ContentOnDemandController is only accessible from inside the cluster. If a hacker gets access, the main problem is the data the hacker is able to get in return - not the url that is sent to puppeteer.

Reply:
Is it unnecessary to inldude QueryString?

Reply:
No, you're right. It's removed now.